netket/_vmc_driver.py

```diff
@@ -125,33 +126,42 @@ def __init__(
         self._machine = sampler.machine
         self._sampler = sampler
         self._sr = sr
+        self._stats = None
 
         self._optimizer_step, self._optimizer_desc = make_optimizer_fn(
             optimizer, self._machine
         )
 
         self._npar = self._machine.n_par
-        self._mc_data = None
 
         if n_samples <= 0:
             raise ValueError(
                 "Invalid number of samples: n_samples={}".format(n_samples)
             )
-        if n_discard is not None and n_discard <= 0:
+        if n_discard is not None and n_discard < 0:
             raise ValueError(
                 "Invalid number of discarded samples: n_discard={}".format(n_discard)
             )
 
-        self.n_samples = n_samples
+        self._n_chains = sampler.n_chains
+
+        self._n_samples = int(self._n_chains * (n_samples // self._n_chains))
```

Comment:
Is there a reason `self._n_chains` is now marked as private with the underscore, while `self.n_discard` is not?

Reply:
good catch, I have now moved n_samples and n_discard to be properties with dedicated setters. I think it's good to leave the freedom to be able to change the number of samples/discarded samples without the need to re-construct the vmc object!

Reply:
It's good to have those values as properties, agreed.